branches/rd235/vde-2/man/vde_vxlan.1
.TH VDE_VXLAN 1 "October 26, 2013" "Virtual Distributed Ethernet"
.SH NAME
vde_vxlan \- VXLAN tunnel endpoint for VDE
.SH SYNOPSIS
.B vde_vxlan
.BI --vxlan-id
.I vxlan_id
.BI --vxlan-addr
.I multicast_address
[
.BI --vxlan-port
.I udp_port
]
[
.BI --vxlan-mttl
.I multicast_ttl
]
[
.BI --sock
.I vde_socket_dir
]
[
.BI --port
.I vde_port
]
[
.BI --daemon
]
[
.BI --verbose
]
.br
.SH DESCRIPTION
A
\fBvde_vxlan\fP 
is a Virtual eXtensible LAN (VXLAN) tunnel endpoint for the vde architecture,
that can connect a \fBvde_switch (1)\fP to a VXLAN segment.

.SH OPTIONS

.TP
.B --vxlan-id
ID of the VXLAN (VNI)
.TP
.B --vxlan-addr
Multicast address of the VXLAN
.TP
.B --vxlan-port
Port of the VXLAN (default 4879)
.TP
.B --vxlan-mttl
Multicast TTL (default 1)
.TP
.B --sock
Socket directory of the VDE switch
.TP
.B --port
Port of the VDE switch
.TP
.B --daemon
Run in background
.TP
.B --verbose
Show debug output

.SH NOTICE
Virtual Distributed Ethernet is not related in any way with
www.vde.com ("Verband der Elektrotechnik, Elektronik und Informationstechnik"
i.e. the German "Association for Electrical, Electronic & Information
Technologies").

.SH SEE ALSO
\fBvde_switch\fP(1),
.br
.SH AUTHORS
vde_vxlan is a tool by Alessandro Ghedini <alessandro@ghedini.me>

VDE is a project by Renzo Davoli <renzo@cs.unibo.it>


Changelog
VDE 2.3.1, 2010-07-25
	* [vde_switch] Fix control socket path resolution regression.

VDE 2.3.0, 2010-05-29
	* Remove some hardcoded paths and use SYSCONFDIR/LOCALSTATEDIR instead.
	* [vde_switch] Fix multiple vlan tagging when packetfilter is used.
	* Build system fixes.
	* Ship README and patch to add VDE support in VirtualBox (included in
	  VirtualBox since 3.2.0).
	* Ship an updated version of slirp taken from QEMU and include VDE support.
	* [vde_cryptcab] Remove SIGALRM-based timers.
	* [vde_l3] Fix unicast test (sourceforge: #2726725)
	* [libvdeplug] Add vdestream abstraction to send/receive bytes (previously in
	  vde_plug)
	* [vde_switch] New command "port/createauto" to create a new port with an
	  automatically allocated ID

VDE 2.2.3, 2009-05-05
	* [vdeterm] Moved history management in a separate library.
	* [vde_switch] Console bug fixes.
	* Added vde_pcapplug (connect a vde_switch to a real interface using pcap)
	* Added missing includes of limits.h/ctype.h
	* [vde_switch] Added option --mgmtgroup to consmgmt module to specify the
	  group of the management socket.
	* [vde_switch] Added option --dirmode to datasock module to specify the mode
	  of the socket directory.
	* [vdetaplib] Added specification of port, mode and group.
	* [kvde_switch] Bugfixes.
	* [vde_switch] pdump plugin: add fifo support and buffered/unbuffered choice.
	* [common] Fix '#include <limits.h>' for PATH_MAX. (Closes: #2023790)
	* [vdetaplib] Manpage: added an example to override default port, group and
	  mode. (Closes: #2066885)
	* [vde_switch] Resize untagged ports bitarray while changing number of ports
	  (Thanks to Michael Sallaway). Closes: #2115547.
	* [vde_switch] Fix memory reset in BA_REALLOC and BAC_REALLOC (by rd235).
	  (Closes: #2123601)
	* [kvde_switch] By default kvde_switch is now not compiled in.
	* [vdeq] Fixed loop if an unknow parameter is specified. Thanks to tfero.
	  (Closes: #2134438)
	* [slirpvde] Imported patch from <ludwig.nussel@suse.de> to fix static buffer
	  overflow (Closes: #2138410).
	* [vde_switch] Fixed compatibility with armel architecture using "attribute
	  packet" on registration message structures (thanks to Michele Cucchi).
	* [vde_cryptcap] Some modifications to allow the client to work on the
	  OpenMoko freerunner.
	* [vde_switch] Added port access control.
	* [vde_switch] Changed regular user standard switch to ~/.vde2/stdsock.
	* [libvdeplug] Downsize vde_open_real() and vde_realpath() stack to fit in
	  User-Mode Linux.
	* [common] Add pkg-config files for VDE libraries.
	* [common] Add '--disable-pcap' configure option, thanks to Wulf C. Krueger
	  (Closes: #2600817).
	* [common] Fix bad malloc replacement (Closes: #2631581), thanks to Michele
	  Cucchi.
	* [vde_cryptcab] Added support for daemonization.
	* [vde_cryptcab] Fixed server side disconnections (thanks to Brendan Grieve).
	* [wirefilter] Added rc file.

VDE 2.2.2, 2008-07-08
	* Added --enable-profile to configure (for profiling)
	* Added cleanup() in vdeq on successfull termination (thanks huslu)
	* Added --mod to new syntax in vdeq man page
	* Fixed fd leak in runscript() (Closes: #2009311 on SF)
	* wirefilter man and code cleanup
	* wirefilter option for blinking added.
	* Gaussian distribution for range values, Gilbert bursty model for packet loss.
	* Bugfix on mgmt sessions.
	* Poor's man hub implemented. (patch by Luca Saiu)

VDE 2.2.1, 2008-06-17
	* New contact email address: <info@v2.cs.unibo.it>
	* Code cleanup (some more warnings removed)
	* [vde_crytcab] Cleared large memory leak due to missing
	  EVP_CIPHER_CTX_cleanup() after encryption/decryption
	* [vde_cryptcab] Cleared (very small) memory leak in crc32 calls
	* [vde_cryptcab] Fixed bug on handover after session timeout: now client
	  re-establishes the connection when it sees some traffic, after session
	  timeout.
	* Heavy directories reorganization, all source code moved under src and
	  include
	* New internal library libvdecommon.la with common utilities and LIBOBJ
	  replacement/compatibility functions
	* [common] Fix in poll emulation management
	* Fix to distclean problem due to wrong LIBOBJ management
	* [vde_switch] hash/setsize 0 was not recognized as an error and caused vde_switch
	  to hang due to an infinite loop
	* Don't compile cryptcab if configure detects not to.
	* Fix for autotools in Slackware
	* [vde_cryptcab] refactoring:
	  - simplified protocol (though keeping it back-compatible)
	  - fixed some bugs and memory leaks
	  - added some new features
	* [vde_switch] fix on BPDU generation/handling: do not generate BPDU if FSTP is not active
	  or we are HUB, forward BPDU if FSTP is not active (Closes #1943973 on SF).
	* [slirpvde] include limits.h on linux as well.
	* [vde_switch] Applied patch from Bjorn Mork to remove 32/64 bit confusion from bitarray.h
	  (Closes #1984460 on SF).
	* [vde_switch] do not update last_seen in hash entry if data arrives from
	  a new port and entry still contains old port (Closes #469098 debian).
	* [vde_cryptcab] server side:
	  - Added feature: cleanup key files on exit
	  - Bugfix: avoid using several switch ports for the same client
	* [vde_switch][libvdeplug] fixed path management: as paths are sent between
	  different processes with potentially different CWDs, we must assure that the
	  exchanged paths are always absolute ones.
	  - added vde_realpath (similar to realpath(3) but with the resulting path
	    well-defined also in case of partial resolution), taken from xmview's
	    canonicalize. vde_realpath is now part of libvdecommon.
	  - [vde_switch, kvde_switch] socket name in datasock gets canonicalized
	  - [vde_switch, kvde_switch] restored CWD of the switch after each module
	    initialization (so each module initializes with the same CWD. consmgmt
	    called daemon(0, 0) which changed the CWD to /, so the behavior was
	    different when the switch ran in background or in foreground.
	  - [vde_switch] clarified some error messages

branches/rd235/vde-2/src/vde_vxlan/vde_vxlan.c
/*
 * VDE - vde_vxlan Network emulator for vde
 * Copyright (C) 2014 Renzo Davoli, Alessandro Ghedini VirtualSquare
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdlib.h>
#include <getopt.h>
#include <poll.h>
#include <unistd.h>

#include "vxlan_hash.h"
#include "log.h"
#include "vxlan.h"
#include "plug.h"

/* from vde_switch/switch.h */

#define INIT_HASH_SIZE 128

void cleanup();
void help();

int main(int argc, char *argv[]) {
	int opts;
	struct pollfd pfd[3];

	char *plug_addr  = NULL;
	int   plug_port  = 0;

	int daemonize = 0;

	const char   *short_opts  = "I:A:P:T:s:p:dvh";
	struct option long_opts[] = {
		{ "vxlan-id",   required_argument, 0, 'I' },
		{ "vxlan-addr", required_argument, 0, 'A' },
		{ "vxlan-port", required_argument, 0, 'P' },
		{ "vxlan-mttl", required_argument, 0, 'T' },

		{ "sock",       required_argument, 0, 's' },
		{ "port",       required_argument, 0, 'p' },

		{ "daemon",     no_argument,       0, 'd' },
		{ "verbose",    no_argument,       0, 'v' },

		{ "help",       no_argument,       0, 'h' },
		{0, 0, 0, 0}
	};

	while ((opts = getopt_long(argc, argv, short_opts, long_opts, 0)) != -1) {
		switch (opts) {
			/* VXLAN opts */
			case 'I': { vxlan_id   = atoi(optarg);      break; }
			/* TODO: IPv6 support */
			case 'A': { vxlan_addr = inet_addr(optarg); break; }
			case 'P': { vxlan_port = atoi(optarg);      break; }
			case 'T': { vxlan_mttl = atoi(optarg);      break; }

			/* VDE opts */
			case 's': { plug_addr = strdup(optarg);     break; }
			case 'p': { plug_port = atoi(optarg);       break; }

			case 'd': { daemonize = 1;                  break; }
			case 'v': { debug = 1;                      break; }

			default :
			case 'h': { help(); exit(1);                       }
		}
	}

	if (vxlan_id == -1) {
		printlog(LOG_ERR, "Invalid VXLAN ID");
		exit(1);
	}

	if (vxlan_addr == INADDR_NONE) {
		printlog(LOG_ERR, "Invalid VXLAN multicast address");
		exit(1);
	}

	atexit(cleanup);

	plug_open(plug_addr, plug_port, pfd);

	vxlan_open(pfd);

	hash_init(INIT_HASH_SIZE);

	if (daemonize && daemon(0, 0)) {
		printlog(LOG_ERR, "daemon(): %s", strerror(errno));
		return 1;
	} else if (daemonize) {
		logok = 1;
		openlog("vde_vxlan", LOG_PID, 0);
		printlog(LOG_INFO, "VDE_VXLAN started");
	}

	while (1) {
		int n = poll(pfd, 3, 1000);

		if ((n < 0) && (errno != EINTR)) {
			printlog(LOG_ERR, "poll(): %s", strerror(errno));
			return 1;
		}

		if (pfd[0].revents & POLLHUP) {
			printlog(LOG_INFO, "VDE connection closed");
			return 0;
		}

		if (pfd[1].revents & POLLHUP) {
			printlog(LOG_ERR, "VDE connection error");
		}

		if (pfd[1].revents & POLLIN) {
			plug_process();
		}

		if (pfd[2].revents & POLLIN) {
			vxlan_process();
		}

		hash_gc();
	}

	return 0;
}

void cleanup() {
	vxlan_close();
	plug_close();
}

void help() {
	#define CMD_HELP(CMDL, CMDS, MSG) printf("  %s, %s\t%s.\n", CMDS, CMDL, MSG);

	puts("Usage: vde_vxlan [OPTIONS]\n");
	puts(" VXLAN Options:");

	CMD_HELP("--vxlan-id",   "-I", "ID of the VXLAN");
	CMD_HELP("--vxlan-addr", "-A", "Multicast address of the VXLAN");
	CMD_HELP("--vxlan-port", "-P", "Port of the VXLAN (default 4879)");
	CMD_HELP("--vxlan-mttl", "-T", "Multicast TTL (default 1)");

	puts("\n VDE Options:");

	CMD_HELP("--sock", "-s", "Socket directory of the VDE switch");
	CMD_HELP("--port", "-p", "Port of the VDE switch");

	CMD_HELP("--daemon",  "-d", "Run in background");
	CMD_HELP("--verbose", "-v", "Show debug output");

	CMD_HELP("--help",    "-h", "Show this help");

	puts("");
}
